src/validation-engine.ts

```diff
@@ -0,0 +1,228 @@
+import {
+  ValidationResult,
+  FormValidationResult,
+  FieldsValidationSchema,
+  RecordValidationFunction,
+  createDefaultValidationResult,
+  RecordValidationFunctionSyncAsync,
+  isSyncValidationResult,
+  FieldValidationFunctionSyncAsync,
+  FieldValidationFunction,
+} from './model';
+
+import { isUndefinedOrNull } from './helper';
+
+import {
+  fireAllFieldsValidations,
+  fireSingleFieldValidations,
+  fireRecordValidations,
+} from './validation-dispatcher';
+import { buildFormValidationResult } from './form-validation-summary-builder';
+
+export class ValidationEngine {
+  private validationsPerField: FieldsValidationSchema = {};
+  private validationsGlobalForm: RecordValidationFunction[] = [];
+  private asyncValidationInProgressCount = 0;
+
+  // TODO, in new api we are adding an optional parameter
+  // to customize the error message (string array)
```

Comment:
Mmm... is this neede here or was it at validator definition level?

Reply:
That's right. We have to update `FieldValidationFunction` type to include the `message` property. isn't it?

```diff
export type FieldValidationFunction = (
  value: any,
  values?: any,
  customParams?: any,
+ message: string | string[];
) => Promise<ValidationResult>;

export interface FullFieldValidation {
  validator: FieldValidationFunction;
  customArgs?: object;
- errorMessage: string | string[];
+ message: string | string[];
}

```

Reply:
Should be optional I guess message?

Reply:
Yes, sure

Reply:
building the test realized that customArgs should be object and customMessage string, by doing this if by mistake you forget to add empty customrArgs typescript will complain and you will just read the signature and realize you have to enter an empty custom args, what do you think?

Reply:
And relized typo on Validation function def it says customParams should be customArgs

Reply:
So maybe we need to create an Interface + object with optional parameters? And it feeds the `validator` with these params. Because, if we want to use value and message, we have to provide values and customArgs?.

What do you think?

Reply:
Field is handling now customMessage, about form validation need to make a refactor the validation schema only holds functions it should be an structure.

Reply:
Done custom message are now supported on both field and record validations